File: app/src/test/java/com/beautifulquran/ui/rootviewer/DictionaryTextTest.kt
package com.beautifulquran.ui.rootviewer

import com.beautifulquran.data.model.DictionaryEntry
import com.beautifulquran.data.model.DictionarySenseGroup
import com.beautifulquran.data.parseDictionaryPayload
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class DictionaryTextTest {

    @Test
    fun `parsePayload reads POS groups`() {
        val groups = parseDictionaryPayload(
            """[{"pos":"verb","glosses":["to say","to tell"]},{"pos":"noun","glosses":["speech"]}]""",
        )
        assertEquals(2, groups.size)
        assertEquals("verb", groups[0].pos)
        assertEquals(listOf("to say", "to tell"), groups[0].glosses)
        assertEquals("noun", groups[1].pos)
    }

    @Test
    fun `parsePayload tolerates spaced JSON from the packed DB`() {
        // build_dictionary_db emits spaces after ':' — must parse on Android ICU too.
        val groups = parseDictionaryPayload(
            """[{"pos": "noun", "glosses": ["fire", "conflagration", "gunfire"]}]""",
        )
        assertEquals(1, groups.size)
        assertEquals("noun", groups[0].pos)
        assertEquals(listOf("fire", "conflagration", "gunfire"), groups[0].glosses)
    }

    @Test
    fun `parsePayload unescapes quoted glosses`() {
        val groups = parseDictionaryPayload(
            """[{"pos":"noun","glosses":["a \"quoted\" sense","line\nbreak"]}]""",
        )
        assertEquals(listOf("a \"quoted\" sense", "line\nbreak"), groups[0].glosses)
    }

    @Test
    fun `preferred QAC POS comes first`() {
        val entry = DictionaryEntry(
            lemma = "قَالَ",
            word = "قال",
            groups = listOf(
                DictionarySenseGroup("noun", listOf("speech")),
                DictionarySenseGroup("verb", listOf("to say")),
            ),
            credit = "credit",
        )
        val glosses = dictionaryGlosses(entry, "V")
        assertEquals("Verb", glosses.first().first)
        assertEquals("to say", glosses.first().second)
    }

    @Test
    fun `preview threshold matches Classical-lexicon style expand`() {
        assertFalse(dictionaryNeedsExpand(DICTIONARY_PREVIEW_SENSES))
        assertTrue(dictionaryNeedsExpand(DICTIONARY_PREVIEW_SENSES + 1))
    }

    @Test
    fun `wiktionary URL encodes the Arabic headword`() {
        assertTrue(wiktionaryArabicUrl("كتاب").contains("كتاب") ||
            wiktionaryArabicUrl("كتاب").contains("%D9%83"))
        assertTrue(wiktionaryArabicUrl("كتاب").endsWith("#Arabic"))
    }
}

File: web/src/ui/root/__tests__/dictionaryText.test.ts
import { describe, expect, it } from 'vitest'
import { parseDictionaryPayload } from '../../../data/dictionary'
import {
  DICTIONARY_PREVIEW_SENSES,
  dictionaryGlosses,
  dictionaryNeedsExpand,
  orderedDictionaryGroups,
  wiktionaryArabicUrl,
} from '../dictionaryText'

describe('parseDictionaryPayload', () => {
  it('reads POS groups', () => {
    const groups = parseDictionaryPayload(
      JSON.stringify([
        { pos: 'verb', glosses: ['to say', 'to tell'] },
        { pos: 'noun', glosses: ['speech'] },
      ]),
    )
    expect(groups).toHaveLength(2)
    expect(groups[0]).toEqual({ pos: 'verb', glosses: ['to say', 'to tell'] })
  })
})

describe('dictionaryGlosses', () => {
  it('puts the preferred QAC POS first', () => {
    const rows = dictionaryGlosses(
      {
        lemma: 'قَالَ',
        word: 'قال',
        groups: [
          { pos: 'noun', glosses: ['speech'] },
          { pos: 'verb', glosses: ['to say'] },
        ],
        credit: 'credit',
      },
      'V',
    )
    expect(rows[0]).toEqual({ pos: 'Verb', gloss: 'to say' })
  })

  it('keeps source order when QAC POS is unknown', () => {
    const groups = orderedDictionaryGroups(
      [
        { pos: 'noun', glosses: ['speech'] },
        { pos: 'verb', glosses: ['to say'] },
      ],
      'XYZ',
    )
    expect(groups.map((g) => g.pos)).toEqual(['noun', 'verb'])
  })
})

describe('dictionary preview', () => {
  it('expands past the preview budget', () => {
    expect(dictionaryNeedsExpand(DICTIONARY_PREVIEW_SENSES)).toBe(false)
    expect(dictionaryNeedsExpand(DICTIONARY_PREVIEW_SENSES + 1)).toBe(true)
  })
})

describe('wiktionaryArabicUrl', () => {
  it('points at the Arabic section', () => {
    expect(wiktionaryArabicUrl('كتاب').endsWith('#Arabic')).toBe(true)
  })
})
